src/gennylib/src/Orchestrator.cpp

```diff
@@ -160,4 +160,12 @@ void Orchestrator::abort() {
     _phaseChange.notify_all();
 }
 
+void Orchestrator::sleepToPhaseEnd(const PhaseNumber pn, Duration timeout) {
+    reader lock{_mutex};
+    if (this->_current != pn || state == State::PhaseEnded) {
+        return;
+    }
+    _phaseChange.wait_for(lock, timeout);
```

Comment:
spurious wakeup could cause this to return early. I think you need to wrap this in a `while` with the inverted `if` predicate above being your loop predicate

```c++
while(this->_current == pn && state != PhaseEnded .....) { _phaseChange.wait_for(...) }
```

Reply:
That's a good point. I originally ignored spurious wakeups because we do this check outside of the orchestrator, but you made me realize that if we don't account for the wakeups then we might not hit the correct rate (since we always sleep for the rate, if every thread woke up right before it was time to burst, then was put back to sleep for the rate again, then we'd go under the rate).

Added your suggestion and some timing-based extra details.